### PR TITLE
Switch to batch/v1 in cronjob's e2es

### PR DESCRIPTION
**Pull request by soltysh**
#### What type of PR is this?

/kind feature
/kind design
/area conformance
/sig testing
/sig apps 
/priority backlog

#### What this PR does / why we need it:
This promotes CronJob e2es to conformance as complimentary to https://github.com/kubernetes/kubernetes/pull/98965.
See kubernetes/enhancements#19 and complete description from https://github.com/kubernetes/enhancements/tree/master/keps/sig-apps/19-Graduate-CronJob-to-Stable

#### Special notes for your reviewer:
/assign @johnbelamaric 

#### Does this PR introduce a user-facing change?
```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:
```docs
- [KEP]: https://github.com/kubernetes/enhancements/tree/master/keps/sig-apps/19-Graduate-CronJob-to-Stable 
```

### CODE REVIEW

**Comment by dims** [comment]
@soltysh curious .. i see an old issue about improving e2e test coverage here https://github.com/kubernetes/kubernetes/issues/85974 

Does the promotion in this PR add any additional coverage? (i do understand that the issue was opened against older version of cronjob)

I was also peeking at https://apisnoop.cncf.io/1.21.0/alpha/batch trying to understand

**Comment by soltysh** [comment]
> Does the promotion in this PR add any additional coverage? (i do understand that the issue was opened against older version of cronjob)
> 
> I was also peeking at https://apisnoop.cncf.io/1.21.0/alpha/batch trying to understand

@dims yes, the above are just promoting pre-existing tests which cover the core CronJob functionality. If you look through it mentions several new tests that I'm just working on. Glad you ask :smile:

**Comment by dims** [comment]
@soltysh thanks for confirming. Looks like https://github.com/kubernetes/kubernetes/pull/98965 need to be merged first and then the `newTestCronJob` needs to be updated to use the GA version of the resource before we can merge this. am i reading this right? (trying to learn the ropes here)

**Comment by soltysh** [comment]
> @soltysh thanks for confirming. Looks like #98965 need to be merged first and then the `newTestCronJob` needs to be updated to use the GA version of the resource before we can merge this. am i reading this right? (trying to learn the ropes here)

I was debating whether these tests should be promoted right away or only in the next release. The reason for that is that only 1.21 has the `batch/v1` CronJob, and conformance have specific requirements for version skew (https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md#conformance-test-version-skew-policy)

**Comment by johnbelamaric** [comment]
> I was debating whether these tests should be promoted right away or only in the next release. The reason for that is that only 1.21 has the `batch/v1` CronJob, and conformance have specific requirements for version skew 

Old versions will not be conformant (necessarily) with new conformance tests. But new versions should be conformant by the standards of older conformance tests.

In other words, your tests should be promoted in the same release as the API.

**Comment by dims** [comment]
interesting... thanks @soltysh

**Comment by soltysh** [comment]
> In other words, your tests should be promoted in the same release as the API.

Sounds good, I'll wait for https://github.com/kubernetes/kubernetes/pull/98965 and https://github.com/kubernetes/kubernetes/pull/99314 and will bump tests to use `batch/v1` in here.

**Comment by soltysh** [comment]
This builds on top of https://github.com/kubernetes/kubernetes/pull/98965 so only the last 2 commits matter, where:
1. moves all cronjob e2e to use `batch/v1`
2. promotes the main bits of cronjob tests as conformance

File: test/e2e/apps/cronjob.go [hunk 1]
```diff
@@ -412,17 +411,33 @@ var _ = SIGDescribe("CronJob", func() {
...
+		framework.ExpectNoError(err)
+		err = cjClient.Delete(context.TODO(), forRemovalCronJob.Name, metav1.DeleteOptions{})
+		framework.ExpectNoError(err)
+		cjs, err = cjClient.List(context.TODO(), metav1.ListOptions{LabelSelector: "special-label=" + f.UniqueName})
+		framework.ExpectNoError(err)
+		// Should have <= 2 items since some cronjobs might not have been deleted yet due to finalizers
+		framework.ExpectEqual(len(cjs.Items) <= 2, true, "filtered list should be <= 2")
+		// Validate finalizers
+		for _, cj := range cjs.Items {
+			expectFinalizer(&cj, "deleting cronjob collection")
```
Comment (dims) [inline test/e2e/apps/cronjob.go:454]: this is just `"deleting cronjob"` right? (will be confused with the one from like 443 otherwise?)

Comment (soltysh) [reply test/e2e/apps/cronjob.go:454]: Yeah, lemme update that quickly right away.

**Comment by dims** [comment]
@soltysh thanks! changes look good now!

**Comment by alaypatel07** [comment]
+1, LGTM

Many thanks, @soltysh !!

File: test/e2e/apps/cronjob.go [hunk 2]
```diff
@@ -412,17 +434,32 @@ var _ = SIGDescribe("CronJob", func() {
...
+
+		ginkgo.By("deleting a collection")
 		err = cjClient.DeleteCollection(context.TODO(), metav1.DeleteOptions{}, metav1.ListOptions{LabelSelector: "special-label=" + f.UniqueName})
 		framework.ExpectNoError(err)
 		cjs, err = cjClient.List(context.TODO(), metav1.ListOptions{LabelSelector: "special-label=" + f.UniqueName})
 		framework.ExpectNoError(err)
-		// Should have <= 1 items since some cronjobs might not have been deleted yet due to finalizers
-		framework.ExpectEqual(len(cjs.Items) <= 1, true, "filtered list should be <= 1")
+		// Should have <= 2 items since some cronjobs might not have been deleted yet due to finalizers
+		framework.ExpectEqual(len(cjs.Items) <= 2, true, "filtered list should be <= 2")
```
Comment (liggitt) [inline test/e2e/apps/cronjob.go:462]: why the change from 1 to 2?

Comment (liggitt) [reply test/e2e/apps/cronjob.go:462]: This PR promotes the test to conformance, but does not include the change to make the controller use the v1 API?

This test passed in the ga-only CI job, which turns off beta APIs. If the controller is still using v1beta1 in this branch, does the fact that the test passed mean this test does not check anything the controller does? Wouldn't we expect to verify basic functionality of cronjob from a user's perspective?

Comment (soltysh) [reply test/e2e/apps/cronjob.go:462]: > why the change from 1 to 2?

This test previously was missing single delete, there was only delete collection. I've added that one bit to have a full API operation coverage. This means that during check there might be 2 cronjobs, so I'm checking for up to 2 cronjobs marked for deletion. One from a single delete, the other from delete collection. 

> This test passed in the ga-only CI job, which turns off beta APIs. If the controller is still using v1beta1 in this branch, does the fact that the test passed mean this test does not check anything the controller does? Wouldn't we expect to verify basic functionality of cronjob from a user's perspective?

lemme rebase this on top of both api and controller changes to ensure this works as expected.

Comment (soltysh) [reply test/e2e/apps/cronjob.go:462]: Also I don't expect this to fail in any way, aside from my personal testing, all of these tests (with the exception of that last change) are part of a regular e2e (non-conformance) for quite a while now and didn't had issues at least in the past weeks.

Comment (liggitt) [reply test/e2e/apps/cronjob.go:462]: > This test previously was missing single delete, there was only delete collection. I've added that one bit to have a full API operation coverage.

ah, got it

**Comment by liggitt** [comment]
lgtm, https://github.com/kubernetes/kubernetes/pull/99423 is in the merge pool

once it merges, rebase this and split it into two PRs:
* one that switches the e2e to use v1
* one that promotes the test to conformance

we'll merge the "switch e2e to v1" PR and let the "promote to conformance" PR soak for two weeks before merging

**Comment by soltysh** [comment]
> * one that switches the e2e to use v1

This was left here as is.

> * one that promotes the test to conformance

Split into https://github.com/kubernetes/kubernetes/pull/99851 which still builds on top of this.
